File: solutions/python/balanced_binary_tree.py
# Balanced_binary_tree.cpp 100b4adabfd008775520062bd407c3323ea646af
from binary_tree_prototype import BinaryTreeNode


# @include
def is_balanced_binary_tree(tree):
    return check_balanced(tree)[0]


# First value of the return value indicates if tree is balanced, if
# balanced the second value of the return value is the height of tree.
def check_balanced(tree):
    if tree is None:
        return (True, -1)  # Base case.

    left_result = check_balanced(tree.left)
    if not left_result[0]:
        return (False, 0)  # Left subtree is not balanced.

    right_result = check_balanced(tree.right)
    if not right_result[0]:
        return (False, 0)  # Right subtree is not balanced.

    is_balanced = abs(left_result[1] - right_result[1]) <= 1
    height = max(left_result[1], right_result[1]) + 1
    return (is_balanced, height)
# @exclude


def main():
    #  balanced binary tree test
    #      3
    #    2   5
    #  1    4 6
    tree = BinaryTreeNode()
    tree.left = BinaryTreeNode()
    tree.left.left = BinaryTreeNode()
    tree.right = BinaryTreeNode()
    tree.right.left = BinaryTreeNode()
    tree.right.right = BinaryTreeNode()
    assert is_balanced_binary_tree(tree) == True
    print(is_balanced_binary_tree(tree))
    # Non-balanced binary tree test.
    tree = BinaryTreeNode()
    tree.left = BinaryTreeNode()
    tree.left.left = BinaryTreeNode()
    assert is_balanced_binary_tree(tree) == False
    print(is_balanced_binary_tree(tree))


if __name__ == '__main__':
    main()

File: solutions/cpp/Balanced_binary_tree.cc
// Copyright (c) 2015 Elements of Programming Interviews. All rights reserved.

#include <algorithm>
#include <cassert>
#include <iostream>
#include <memory>
#include <utility>

#include "./Binary_tree_prototype.h"

using std::boolalpha;
using std::cout;
using std::endl;
using std::make_unique;
using std::max;
using std::pair;
using std::tie;
using std::unique_ptr;

pair<bool, int> CheckBalanced(const unique_ptr<BinaryTreeNode<int>>&);

// @include
bool IsBalanced(const unique_ptr<BinaryTreeNode<int>>& tree) {
  return CheckBalanced(tree).first;
}

// First value of the return value indicates if tree is balanced, and if
// balanced the second value of the return value is the height of tree.
pair<bool, int> CheckBalanced(const unique_ptr<BinaryTreeNode<int>>& tree) {
  if (tree == nullptr) {
    return {true, -1};  // Base case.
  }

  bool is_left_balanced;
  int left_height;
  tie(is_left_balanced, left_height) = CheckBalanced(tree->left);
  if (!is_left_balanced) {
    return {false, 0};  // Left subtree is not balanced.
  }
  bool is_right_balanced;
  int right_height;
  tie(is_right_balanced, right_height) = CheckBalanced(tree->right);
  if (!is_right_balanced) {
    return {false, 0};  // Right subtree is not balanced.
  }

  bool is_balanced = abs(left_height - right_height) <= 1;
  int height = max(left_height, right_height) + 1;
  return {is_balanced, height};
}
// @exclude

int main(int argc, char* argv[]) {
  //  balanced binary tree test
  //      3
  //    2   5
  //  1    4 6
  unique_ptr<BinaryTreeNode<int>> tree =
      make_unique<BinaryTreeNode<int>>(BinaryTreeNode<int>());
  tree->left = make_unique<BinaryTreeNode<int>>(BinaryTreeNode<int>());
  tree->left->left = make_unique<BinaryTreeNode<int>>(BinaryTreeNode<int>());
  tree->right = make_unique<BinaryTreeNode<int>>(BinaryTreeNode<int>());
  tree->right->left = make_unique<BinaryTreeNode<int>>(BinaryTreeNode<int>());
  tree->right->right = make_unique<BinaryTreeNode<int>>(BinaryTreeNode<int>());
  assert(IsBalanced(tree) == true);
  cout << boolalpha << IsBalanced(tree) << endl;
  // Non-balanced binary tree test.
  tree = make_unique<BinaryTreeNode<int>>(BinaryTreeNode<int>());
  tree->left = make_unique<BinaryTreeNode<int>>(BinaryTreeNode<int>());
  tree->left->left = make_unique<BinaryTreeNode<int>>(BinaryTreeNode<int>());
  assert(IsBalanced(tree) == false);
  cout << boolalpha << IsBalanced(tree) << endl;
  return 0;
}
